### PR TITLE
fixes bug 1408448 - stomp on signature when parentBuildID != childBuildID

### PR DESCRIPTION
**Pull request by willkg**
If `MozCrashReason` is `"MOZ_RELEASE_ASSERT(parentBuildID == childBuildID)"`, then
stomp on the signature with `"parentBuildID != childBuildID"`. This allows all
these crashes to end up in the same bucket where the original signature is
likely to be bogus since there are two different builds involved.

This also adds `raw_crash['MozCrashReason']` to the list of things that we use to
generate signatures.

I tested this out in this bug comment with the specific crashes and Andrew approved the output.

I think to test, we should let Travis/Circle do their thing. You can also rerun what I did in this comment and verify the output is the same:

https://bugzilla.mozilla.org/show_bug.cgi?id=1408448#c6

### CODE REVIEW

File: socorro/signature/rules.py [hunk 1]
```diff
@@ -710,3 +710,43 @@ def action(self, raw_crash, processed_crash, notes):
             raw_crash['IPCMessageName']
         )
         return True
+
+
+class SignatureIPCMessageName(Rule):
```
Comment (Osmose) [inline socorro/signature/rules.py]: This is a copy-paste of the rule above, maybe you hit paste twice :P

File: socorro/signature/rules.py [hunk 2]
```diff
@@ -710,3 +710,43 @@ def action(self, raw_crash, processed_crash, notes):
...
+        return bool(raw_crash.get('MozCrashReason', '') == value)
+
+    def action(self, raw_crash, processed_crash, notes):
+        notes.append(
+            'Signature replaced with MozCrashAssert, was: "%s"'
+            % processed_crash['signature']
+        )
+
+        # This is a failed assertion and the crash reason is the assertion that failed, so we put
+        # "!=" in the signature because that's what caused the assertion failure.
```
Comment (Osmose) [inline socorro/signature/rules.py]: I think it's clear what the line does without the comment. We can leave it in if you like anyway.

File: socorro/signature/rules.py [hunk 3]
```diff
@@ -710,3 +710,43 @@ def action(self, raw_crash, processed_crash, notes):
...
+
+    In the case where the assertion fails, then the parent buildid and the child buildid are
+    different. This causes a lot of strangeness particularly in symbolification, so the signatures
+    end up as junk. Instead, we want to bucket all these together so we replace the signature.
+
+    """
+
+    def predicate(self, raw_crash, processed_crash):
+        value = 'MOZ_RELEASE_ASSERT(parentBuildID == childBuildID)'
+        return bool(raw_crash.get('MozCrashReason', '') == value)
```
Comment (Osmose) [inline socorro/signature/rules.py]: Is this bool cast necessary?

File: socorro/signature/rules.py (hunk 2 above)
Comment (willkg) [reply socorro/signature/rules.py]: I implemented this wrong the first time. I thought it looked sufficiently curious and wrong that it was prudent to add a verbose comment.

File: socorro/signature/rules.py (hunk 3 above)
Comment (willkg) [reply socorro/signature/rules.py]: You're right--it's already a bool. Oops.

File: socorro/signature/rules.py (hunk 1 above)
Comment (willkg) [reply socorro/signature/rules.py]: I'm having a fucking hell of a time with my keyboard right now. It keeps double-doing things. But I should have caught this before committing.

I'll finish up the work for bug 1377254 (flake8) which would have caught this in CI and local testing.

**Comment by willkg** [comment]
^^^ That rebases against master and then fixes the issues you brought up. This is ready for review again.

**Comment by willkg** [comment]
Thank you!

**Pull request merged**